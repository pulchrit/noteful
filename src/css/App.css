.page-content {
  display: flex;
  padding: 5px;
}

.add-button {
  display: block;
  text-align: center;
  padding: 15px;
  width: 200px;
  border: 1px solid black;
  margin: 20px 10px;
}