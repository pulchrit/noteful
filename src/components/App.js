import React from 'react';
import {Route} from 'react-router-dom';
import NotefulContext from './NotefulContext';
import Header from "./Header";
import MainFolderSidebar from './MainFolderSidebar';
import NoteSidebar from './NoteSidebar';
import MainFolderMain from './MainFolderMain';
import NoteMain from './NoteMain';
import AddNote from './AddNote';
import AddFolder from './AddFolder';
import NotefulError from './NotefulError';
import '../css/App.css';

class App extends React.Component {

  state = {
    folders: [],
    notes: [],
    error: null,
/*     addNote: () => {},
    addFolder: () => {},
    deleteNote: () => {} */
  }

  addNote = (note) => {
    this.setState({
      notes: [ ...this.state.notes, note ]
    })
  }

  addFolder = (folder) => {
    this.setState({
      folders: [ ...this.state.folders, folder ]
    })  
  }

  deleteNote = (noteId) => {
    const newNotes = this.state.notes.filter(note => 
      note.id !== noteId);

    this.setState({
        notes: newNotes
    });
  }

  componentDidMount() {

    // Attribution for Promise.all and getFoldersNotes function: https://tinyurl.com/y42df8dz
    const endpoints = ["http://localhost:9090/folders", "http://localhost:9090/notes"];

    // Promise.all will call this function and fetch the data for each endpoint (folder and
    // notes). It should return the folers/notes as a json object or return an error if the 
    // Promise was rejected.
    const getFoldersNotes = (endpoint) => fetch(endpoint, {
      method: 'GET',
      headers: {
        'content-type': 'application/json',
      }
    })
      .then(res => {
        if (!res.ok) {
          throw new Error(res.status)
        }
        return res.json()
      })
  

    // Call each endpoint, but block the data processing (i.e., setting state) until
    // both Promises have returned successfully or been rejected. (And, actually, I think
    // Promise.all will reject as soon as one of the promises rejects.)
    Promise
      .all(endpoints.map(getFoldersNotes))
      .then(dataArray => 
          this.setState({
              folders: dataArray[0],
              notes: dataArray[1]
          }))
      .catch(error => this.setState({ error }))
    }

  render() {
    // Save contextValue based on state and the methods to update state
    // and pass this contextValue to the Context.Provider below. Now, the 
    // components that render within the ContextProvider will have access to this
    // updated state/context.
    const contextValue = {
      folders: this.state.folders,
      notes: this.state.notes,
      deleteNote: this.deleteNote,
      addFolder: this.addFolder,
      addNote: this.addNote
    }

    return (
      <>
        <Header /> 

        <NotefulContext.Provider value={contextValue}>
          <div className="page-content">

            {/* Sidebar routes */ }
            <NotefulError>
              <Route
                  exact
                  path='/' 
                  component={MainFolderSidebar} 
              />

              <Route
                  path="/folder/:folderId"
                  component={MainFolderSidebar} 
              />

              <Route 
                path='/notes/:noteId'
                component={NoteSidebar}
              />
            </NotefulError>

            {/* Main routes */ }
            <NotefulError>
              <Route  
                exact
                path='/'
                component={MainFolderMain}    
              />

              <Route 
                path="/folder/:folderId"
                component={MainFolderMain}
              />


              <Route 
                path='/notes/:noteId'
                component={NoteMain}
              />
            </NotefulError>

            {/*  Add note and folder routes */ }
            <NotefulError>
              <Route 
                  path='/addNote'
                  component={AddNote}
              />
            </NotefulError>

            <NotefulError>
              <Route 
                  path='/addFolder'
                  component={AddFolder}
              />
            </NotefulError>
          
          </div>
        </NotefulContext.Provider>
        
      </>
    );
  }
}

export default App;
