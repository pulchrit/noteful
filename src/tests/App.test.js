import React from 'react';
import ReactDOM from 'react-dom';
import {BrowserRouter} from 'react-router-dom';
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import App from '../components/App';

describe("App component", () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <BrowserRouter>
        <App />
      </BrowserRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  
  it('renders UI as expected', () => {
    const wrapper = shallow(<App />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

