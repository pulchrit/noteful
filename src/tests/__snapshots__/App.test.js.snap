// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders UI as expected 1`] = `
<Fragment>
  <Header />
  <ContextProvider
    value={
      Object {
        "addFolder": [Function],
        "addNote": [Function],
        "deleteNote": [Function],
        "folders": Array [],
        "notes": Array [],
      }
    }
  >
    <div
      className="page-content"
    >
      <NotefulError>
        <Route
          component={[Function]}
          exact={true}
          path="/"
        />
        <Route
          component={[Function]}
          path="/folder/:folderId"
        />
        <Route
          component={[Function]}
          path="/notes/:noteId"
        />
      </NotefulError>
      <NotefulError>
        <Route
          component={[Function]}
          exact={true}
          path="/"
        />
        <Route
          component={[Function]}
          path="/folder/:folderId"
        />
        <Route
          component={[Function]}
          path="/notes/:noteId"
        />
      </NotefulError>
      <NotefulError>
        <Route
          component={[Function]}
          path="/addNote"
        />
      </NotefulError>
      <NotefulError>
        <Route
          component={[Function]}
          path="/addFolder"
        />
      </NotefulError>
    </div>
  </ContextProvider>
</Fragment>
`;
